File: Coding/python/in_mem_db.py
"""Port of java/InMemDB.java -> in-memory DB with add/delete/get/count + transactions.

Original Java package: code.ds

Faithful port: the reference SET/ROLLBACK count bookkeeping is buggy, so the
GET/COUNT outputs differ from the Java source comments (kept intentionally).
"""

from typing import Dict, List


class MemDB:
    def __init__(self):
        self.map: Dict[str, int] = {}
        self.map1: Dict[int, int] = {}
        self._actions: List["Transaction"] = []

    def SET(self, name: str, value: int) -> None:
        if name in self.map:
            old_value = self.map[name]
            old_count = self.map1[old_value]
            old_count -= 1
            self.map1[old_value] = old_count
            if self._actions:
                self._actions[-1].add_set_command([name, str(old_value)])
        else:
            if self._actions:
                self._actions[-1].add_set_command([name])
        if value not in self.map1:
            self.map1[value] = 1
        else:
            count = self.map1[value]
            count += 1
            self.map1[value] = count
        self.map[name] = value

    def GET(self, name: str) -> None:
        print(self.map.get(name))

    def DELETE(self, name: str) -> None:
        old_value = self.map[name]
        del self.map[name]
        count = self.map1[old_value]
        count -= 1
        self.map1[old_value] = count
        if self._actions:
            self._actions[-1].add_set_command([name, str(old_value)])

    def COUNT(self, value: int) -> None:
        print(self.map1.get(value))

    def BEGIN(self) -> None:
        self._actions.append(Transaction(self))

    def ROLLBACK(self) -> None:
        if not self._actions:
            print("NO TRANSACTIONS IN PROGRESS")
            return
        self._actions.pop().rollback()

    def COMMIT(self) -> None:
        if not self._actions:
            print("NO TRANSACTIONS IN PROGRESS")
            return
        self._actions.clear()


class Transaction:
    def __init__(self, db: MemDB):
        self._db = db
        self._commands: List[str] = []

    def add_set_command(self, command: List[str]) -> None:
        if len(command) == 1:
            self._commands.append(command[0])
        else:
            self._commands.append(command[0] + " " + command[1])

    def add_delete_command(self, command: List[str]) -> None:
        self._commands.append(command[0] + " " + command[1])

    def rollback(self) -> None:
        while self._commands:
            command = self._commands.pop()
            parsed = command.split(" ")
            if len(parsed) == 1:
                self._db.DELETE(parsed[0])
            else:
                self._db.SET(parsed[0], int(parsed[1]))


if __name__ == "__main__":
    db = MemDB()
    db.SET("a", 10)
    db.BEGIN()
    db.SET("a", 15)
    db.BEGIN()
    db.DELETE("a")
    db.SET("a", 10)
    db.ROLLBACK()
    db.COMMIT()
    db.SET("b", 10)
    db.GET("a")    # 15  (Java comment said 10; reference bug preserved)
    db.COUNT(10)   # 1   (Java comment said 2)
    db.SET("a", 10)
    db.GET("a")    # 10
    db.SET("b", 15)
    db.COUNT(15)   # 1
    db.SET("b", 20)
    db.SET("c", 15)
    db.COUNT(10)   # 1

File: Coding/golang/inmemdb/inmemdb.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// MemDB is an in-memory key/value store that supports nested transactions via
// SET, DELETE, GET, COUNT, BEGIN, ROLLBACK and COMMIT.
type MemDB struct {
	m       map[string]int // name -> value
	count   map[int]int    // value -> number of names currently holding it
	actions []*Transaction // open transaction stack
}

func NewMemDB() *MemDB {
	return &MemDB{m: make(map[string]int), count: make(map[int]int)}
}

func (db *MemDB) SET(name string, value int) {
	if oldValue, ok := db.m[name]; ok {
		oldCount := db.count[oldValue]
		oldCount--
		db.count[oldValue] = oldCount
		if len(db.actions) != 0 {
			db.actions[len(db.actions)-1].addSetCommand([]string{name, strconv.Itoa(oldValue)})
		}
	} else {
		if len(db.actions) != 0 {
			db.actions[len(db.actions)-1].addSetCommand([]string{name})
		}
	}
	if _, ok := db.count[value]; !ok {
		db.count[value] = 1
	} else {
		db.count[value]++
	}
	db.m[name] = value
}

func (db *MemDB) GET(name string) {
	if v, ok := db.m[name]; ok {
		fmt.Println(v)
	} else {
		fmt.Println("null")
	}
}

func (db *MemDB) DELETE(name string) {
	oldValue := db.m[name]
	delete(db.m, name)
	db.count[oldValue]--
	if len(db.actions) != 0 {
		db.actions[len(db.actions)-1].addSetCommand([]string{name, strconv.Itoa(oldValue)})
	}
}

func (db *MemDB) COUNT(value int) {
	if c, ok := db.count[value]; ok {
		fmt.Println(c)
	} else {
		fmt.Println("null")
	}
}

func (db *MemDB) BEGIN() {
	db.actions = append(db.actions, NewTransaction(db))
}

func (db *MemDB) ROLLBACK() {
	if len(db.actions) == 0 {
		fmt.Println("NO TRANSACTIONS IN PROGRESS")
		return
	}
	t := db.actions[len(db.actions)-1]
	db.actions = db.actions[:len(db.actions)-1]
	t.rollback()
}

func (db *MemDB) COMMIT() {
	if len(db.actions) == 0 {
		fmt.Println("NO TRANSACTIONS IN PROGRESS")
		return
	}
	db.actions = db.actions[:0]
}

// Transaction records the inverse commands needed to roll back a set of changes.
type Transaction struct {
	commands []string
	db       *MemDB
}

func NewTransaction(db *MemDB) *Transaction {
	return &Transaction{db: db}
}

func (t *Transaction) addSetCommand(command []string) {
	if len(command) == 1 {
		t.commands = append(t.commands, command[0])
	} else {
		t.commands = append(t.commands, command[0]+" "+command[1])
	}
}

func (t *Transaction) addDeleteCommand(command []string) {
	t.commands = append(t.commands, command[0]+" "+command[1])
}

func (t *Transaction) rollback() {
	for len(t.commands) != 0 {
		command := t.commands[len(t.commands)-1]
		t.commands = t.commands[:len(t.commands)-1]
		parsed := strings.Split(command, " ")
		if len(parsed) == 1 {
			t.db.DELETE(parsed[0])
		} else {
			v, _ := strconv.Atoi(parsed[1])
			t.db.SET(parsed[0], v)
		}
	}
}

func main() {
	db := NewMemDB()
	db.SET("a", 10)
	db.BEGIN()
	db.SET("a", 15)
	db.BEGIN()
	db.DELETE("a")
	db.SET("a", 10)
	db.ROLLBACK()
	db.COMMIT()
	db.SET("b", 10)
	db.GET("a")  // 10
	db.COUNT(10) // 2
	db.SET("a", 10)
	db.GET("a")
	db.SET("b", 15)
	db.COUNT(15)
	db.SET("b", 20)
	db.SET("c", 15)
	db.COUNT(10)
}
